src/object/cli_obj.c

```diff
@@ -2622,3 +2660,425 @@ dc_obj_query_key(tse_task_t *api_task)
+					  DIOF_TO_SPEC_SHARD,
+					  unpack_arg->start + i, &io->ui_dkey,
+					  count, iods_o, &sgl, NULL, NULL);
+		if (rc != 0) {
+			D_INFO(DF_OID" (reps %d, failed) when fetch data from "
+			       "shard %d: rc = %d, (2)\n",
+			       DP_OID(unpack_arg->oid), unpack_arg->replicas,
+			       unpack_arg->start + i, rc);
+			D_GOTO(out, rc);
+		}
+
+		if (iovs[0].iov_len != iovs[1].iov_len) {
+			D_INFO(DF_OID" (reps %d, inconsistent) different data "
+			       "length for shard %d (%ld) and shard %d (%ld)\n",
+			       DP_OID(unpack_arg->oid), unpack_arg->replicas,
+			       unpack_arg->start, iovs[0].iov_len,
+			       unpack_arg->start + i, iovs[1].iov_len);
+			D_GOTO(out, rc = -DER_MISMATCH);
+		}
+
+		if (memcmp(iovs[0].iov_buf, iovs[1].iov_buf,
+			   iovs[0].iov_len) != 0) {
+			D_INFO(DF_OID" (reps %d, inconsistent) differnt data "
+			       "for shard %d and shard %d\n",
+			       DP_OID(unpack_arg->oid), unpack_arg->replicas,
+			       unpack_arg->start, unpack_arg->start + i);
+			D_GOTO(out, rc = -DER_MISMATCH);
+		}
+
+		/* Reuse the iovs[1] buffer. Reset it before reusing. */
+		memset(iovs[1].iov_buf, 0, iovs[1].iov_len);
+	}
+
+out:
+	if (iods_o != NULL) {
+		for (i = 0; i < count; i++)
+			daos_iov_free(&iods_o[i].iod_name);
+
+		D_FREE(iods_o);
+	}
+	if (!daos_handle_is_inval(th))
+		dc_tx_local_close(th);
+	for (i = 0; i < 2; i++) {
+		if (iovs[i].iov_buf != NULL)
+			D_FREE(iovs[i].iov_buf);
+	}
+
+	return rc;
+}
+
+static int
+dc_obj_verify_replicas(struct dc_object *obj, unsigned int start,
+		       daos_obj_list_obj_cb_t list_cb,
+		       daos_obj_fetch_cb_t fetch_cb)
+{
+	struct obj_enum_dkeys_arg		*oeda = NULL;
+	struct obj_verify_replicas_unpack_arg	 unpack_arg;
+	daos_epoch_t				 epoch = DAOS_EPOCH_MAX;
+	daos_handle_t				 oh;
+	daos_obj_id_t				 oid;
+	unsigned int				 reps;
+	int					 i = 0;
+	int					 rc = 0;
+
+	oh = obj_ptr2hdl(obj);
+	oid = obj->cob_md.omd_id;
+	reps = obj_get_replicas(obj);
+	D_ALLOC_ARRAY(oeda, reps);
+	if (oeda == NULL) {
+		D_ERROR(DF_OID" no MEM for verify group, start %u, reps %u\n",
+			DP_OID(oid), start, reps);
+		return -DER_NOMEM;
+	}
+
+	for (i = 0; i < reps; i++) {
+		daos_unit_oid_t		shard_id = {
+			.id_pub		= oid,
+			.id_shard	= start + i,
+		};
+
+		daos_enum_dkeys_init_arg(&oeda[i], shard_id);
+	}
+
+	unpack_arg.oh		= oh;
+	unpack_arg.fetch_cb	= fetch_cb;
+	unpack_arg.oid		= oid;
+	unpack_arg.start	= start;
+	unpack_arg.replicas	= reps;
+
+	while (1) {
+		struct obj_enum_dkeys_arg	*arg_l;
+		int				 leader;
+
+leader:
+		leader = obj_grp_leader_get(obj, start, obj->cob_version);
+		if (leader < 0) {
+			D_ERROR(DF_OID" no valid shard, start %u, reps %u: "
+				"rc = %d.\n",
+				DP_OID(oid), start, reps, leader);
+			D_GOTO(out, rc = leader);
+		}
+
+		leader -= start;
+
+list:
+		arg_l = &oeda[leader];
+		rc = daos_enum_dkeys_do_list(oh, &epoch, arg_l, list_cb,
+					     DIOF_TO_SPEC_SHARD |
+					     DIOF_COMMIT_DTX);
+
```

Comment:
Hmm, I am not sure you need that retry logic in dc_enum_dkey_do_list().  TBH, I think this is being overthinked.  IMHO, this tool is just used to verify if the object replication(EC) is consistency.  I do not think you need care about the retry at all, if you might some uncommitted records, just return -EBUSY for now.  I think the normal scenario is that, the tool will somehow open the pool (or container, or object) exclusively, then verify each replicas. i.e. you do not need consider the race between verification and other modification  at all.   I think this will simplified the code a lot.

Reply:
Sorry, just be aware of that how can we know whether there is concurrent modification or not? If the non-leader return different things than from leader, how do you know whether it is caused by concurrent modification or not? Means under which case the logic can return -DER_BUSY?